### PR TITLE
Fix for #7765

### PR DESCRIPTION
**Pull request by stmeyer**
This is a fix for https://github.com/cakephp/cakephp/issues/7765.

Hope I'm doing this well, its my first pull request

### CODE REVIEW

File: src/Database/Dialect/SqlserverDialectTrait.php [hunk 1]
```diff
@@ -325,6 +325,10 @@ public function rollbackSavePointSQL($name)
      */
     public function newCompiler()
     {
+        if (isset($this->_config['compiler'])) {
+            $class = $this->_config['compiler'] . 'Compiler';
```
Comment (jadb) [inline src/Database/Dialect/SqlserverDialectTrait.php]: Maybe throw a detailed exception if the compiler is missing (i.e. missing controllers, etc.)? 

Not sure what others think.

**Comment by jadb** [comment]
:clap: nice job!

Maybe squash into one commit?

File: src/Database/SqlserverWithTriggerCompiler.php [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+ * for SQL Server
+ *
+ * @internal
+ */
+class SqlserverWithTriggerCompiler extends SqlserverCompiler
+{
+    /**
+     * Generates the INSERT part of a SQL query
+     *
+     * SQL Server with enalbed triggers does not allow the OUTPUT clause
```
Comment (markstory) [inline src/Database/SqlserverWithTriggerCompiler.php]: Typo in 'enabled'

File: src/Database/SqlserverWithTriggerCompiler.php [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+     * @param array $parts The parts to build
+     * @param \Cake\Database\Query $query The query that is being compiled
+     * @param \Cake\Database\ValueBinder $generator the placeholder generator to be used in expressions
+     * @return string
+     */
+    protected function _buildInsertPart($parts, $query, $generator)
+    {
+        $table = $parts[0];
+        $talbeParts = explode('.', $table);
+        if (count($talbeParts) > 1) {
```
Comment (markstory) [inline src/Database/SqlserverWithTriggerCompiler.php]: Typo in `tableParts`.

File: src/Database/SqlserverWithTriggerCompiler.php [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+                if (isset($descriptionOfSingleColumn['autoIncrement']) && $descriptionOfSingleColumn['autoIncrement']) {
+                    $descriptionOfSingleColumn['autoIncrement'] = false;
+                }
+                $description->addColumn($columnName, $descriptionOfSingleColumn);
+                $sqlCreateColumns[] = $dialect->columnSql($description, $columnName);
+                $primaryKeyColumnsForSql[] = $columnName;
+            }
+        }
+        $columns = $this->_stringifyExpressions($parts[1], $generator);
+        return sprintf('DECLARE @var TABLE (%s);INSERT INTO %s (%s) OUTPUT INSERTED.%s INTO @var', implode(', ', $sqlCreateColumns), $table, implode(', ', $columns), implode(' , INSERTED.', $primaryKeyColumnsForSql));
```
Comment (markstory) [inline src/Database/SqlserverWithTriggerCompiler.php]: Could this really long line be wrapped?

File: src/Database/SqlserverWithTriggerCompiler.php [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        $driver = $db->driver();
+        $dialect = $driver->schemaDialect();
+        $collection = $db->schemaCollection();
+        $description = $collection->describe($tableWithoutSchema);
+        $constraints = $description->constraints();
+        $primaryKeyColumns = [];
+        foreach ($constraints as $constraintName) {
+            $constraint = $description->constraint($constraintName);
+            if ($constraint['type'] === \Cake\Database\Schema\Table::CONSTRAINT_PRIMARY) {
+                $primaryKeyColumns = $constraint['columns'];
```
Comment (markstory) [inline src/Database/SqlserverWithTriggerCompiler.php]: If you want the primary key columns, the `primaryKey()` method on `Schema\Table` will do just that.

**Comment by markstory** [comment]
Having some tests so we can ensure this works now and in the future would be great. Some tests I would like to see are:

1. Test that inserts and updates work.
2. Interact with schema that has autoincrement primary keys
3. Interact with UUID primary keys
4. Interact with composite primary keys. This might overlap with 2.

**Comment by stmeyer** [comment]
@markstory Where can I find comparable tests?

**Comment by lorenzo** [comment]
@stmeyer what I would do is to create another testsuite using phpunit and run the core database + orm tests after altering the configuration for sqlserver to use this new compiler.

**Comment by stmeyer** [comment]
@lorenzo I never worked with testsuits. I could copy the database testsuit:
```xml
<testsuite name="Database Test Suite">
    <file>./tests/TestCase/DatabaseSuite.php</file>
</testsuite>
```
how to alter the configuration of sqlserver?

**Comment by lorenzo** [comment]
@stmeyer I think I can help you if you give me commit access to your fork, so that I can add commits on top of yours

**Comment by stmeyer** [comment]
@lorenzo I have added you as a collaborator. Thanks for help!

**Comment by stmeyer** [comment]
The tests failed, due to a mistake in TravisCI (https://travis-ci.org/cakephp/cakephp/jobs/111167184#L185):
```
  [Composer\Downloader\TransportException]   
  Could not authenticate against github.com  
```
Is there a way to restart the checks without a new commit?

**Comment by ADmad** [comment]
@stmeyer Restarted

**Comment by stmeyer** [comment]
@ADmad thanks :clap: 
@lorenzo just as information for future planning: when are you planning to create the new testsuit?

**Comment by lorenzo** [comment]
I'm waiting for the weekend to work on this

**Comment by stmeyer** [comment]
@lorenzo nice to hear :smile:

**Comment by lorenzo** [comment]
I need to close this pull request and open it again with the code I added in and after rebasing the branch. It seems like it went crazy

**Pull request closed without merging**